QuoridorEngine/src/core/QuoridorBoard.cs

```diff
@@ -3,184 +3,155 @@
+    /// with stored information.
     /// </summary>
     public class QuoridorBoard
     {
-        List<(int, int)> usedWalls;
-
-        // The complete number of squares per row/column
-        // including the ones destined for walls
         private readonly int dimension;
 
-        public int Dimension { get { return dimension; } }
+        // Using bool arrays to represent used walls
+        // and grid corners for faster search times.
+        private bool[,] corners;
+        private bool[,] verticalWallParts;
+        private bool[,] horizontalWallParts;
+
+        public int Dimension { get; }
 
         public QuoridorBoard(int dimension)
         {
-            usedWalls = new List<(int, int)>();
             this.dimension = dimension;
+
+            corners = new bool[dimension - 1, dimension - 1];
+            InitializeArray(corners);
+
+            verticalWallParts = new bool[dimension, dimension - 1];
+            InitializeArray(verticalWallParts);
+
+            horizontalWallParts = new bool[dimension - 1, dimension];
+            InitializeArray(horizontalWallParts);
         }
 
         /// <summary>
-        /// Utility function to check whether provided 
-        /// square coords represent valid grid square. 
+        /// Utility method to initialize all board arrays
         /// </summary>
-        /// <param name="row">Square row</param>
-        /// <param name="col">Square column</param>
-        /// <returns>True if square is valid else false</returns>
-        public bool IsGridSquare(int row, int col)
+        /// <param name="array">Array to be initialized</param>
+        private static void InitializeArray(bool[,] array)
         {
-            return row >= 2 && col >= 2 && row < dimension && col < dimension && row != col;
-        } 
+            for (int row = 0; row < array.GetLength(0); row++)
+            {
+                for (int col = 0; col < array.GetLength(1); col++)
+                {
+                    array[row, col] = false;
+                }
+            }
+        }
 
         /// <summary>
-        /// Utility function to check whether provided square coords 
-        /// represent player square. If square row or column
-        /// index is odd then it's not a player's square.
+        /// Check if specified horizontal wall part exists
+        /// after necessary coords transformation.
         /// </summary>
-        /// <param name="row">Square row</param>
-        /// <param name="col">Square column</param>
-        /// <returns>True if player's square else false</returns>
-        public bool IsPlayerSquare(int row, int col)
+        /// <param name="row">Wall part row</param>
+        /// <param name="col">Wall part column</param>
+        /// <returns>True if wall part exists</returns>
+        public bool ChechWallPartHorizontal(int row, int col)
         {
-            return row % 2 == 0 && col % 2 == 0;
+            return horizontalWallParts[row - 1, col];
```

Comment:
Maybe add some assertions to verify that given coordinates are inside array bounds. Also applies to other similar functions in this class